### PR TITLE
MV-477: [MV-MIOpen] Benchmark & Port SoftmaxCrossEntropyWithLogits Contiguous

### PR DESCRIPTION
**Pull request by hieule88**
* Added contiguous SoftmaxCrossEntropyWithLogits forward and backward operation and kernel.
* Added driver test and gtest for SoftmaxCrossEntropyWithLogits .
* New API is guarded by MIOPEN_BETA_API macro.
* Compared to ROCm pytorch: [Link](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7B3E71B536-3742-4497-892A-F29002E7C2BC%7D&file=SoftmaxCrossEntropyWithLogits_benchmark.xlsx&action=default&mobileredirect=true&wdsle=0)

* Average over all cases:

|type    |Forward| Backward |
|--------|-------|-------|
|float16 | 3.01| 4.27 |
|float32 | 2.72| 2.99 | 
|bfloat16| 2.99| 4.43|

### CODE REVIEW

File: src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+template <typename TI, typename TO>
+__device__ void softmaxcrossentropywithlogitsForwardContiguous(const TI* __restrict__ input,
+                                                               const TI* __restrict__ target,
+                                                               TO* __restrict__ output,
+                                                               TO* __restrict__ backprop,
+                                                               tensor_view_2d_t input_tv,
+                                                               tensor_view_2d_t target_tv,
+                                                               tensor_view_1d_t output_tv,
+                                                               tensor_view_2d_t backprop_tv)
```
Comment (et16kr) [inline src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp]: Only num_class is used and the remaining tensor_views are not used.

File: src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+{
+    uint64_t gid     = blockIdx.x;
+    uint64_t lid     = threadIdx.x;
+    size_t num_class = input_tv.size[1];
+
+    //   __local FLOAT_ACCUM lmax[LOCAL_SIZE], lsum[LOCAL_SIZE], lloss[LOCAL_SIZE];
```
Comment (et16kr) [inline src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp]: Please remove comment

File: src/solver/softmaxcrossentropywithlogits/fwd_con_softmaxcrossentropywithlogits.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        auto kernel = KernelInfo{};
+
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+            {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+            {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+            {"INPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype},
+            {"OUTPUT_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype},
+            {"D_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype},
```
Comment (et16kr) [inline src/solver/softmaxcrossentropywithlogits/fwd_con_softmaxcrossentropywithlogits.cpp]: D_TYPE is probably not used.

File: test/gtest/softmaxcrossentropywithlogits.hpp [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+        return os << " input:" << tc.input << " contiguous:" << tc.contiguous;
+    }
+
+    std::vector<size_t> GetInput() const { return input; }
+};
+
+inline std::vector<SoftmaxCrossEntropyWithLogitsTestCase> SoftmaxCrossEntropyWithLogitsTestConfigs()
+{
+    return {
+        // {{2000, 3000}, true},
```
Comment (et16kr) [inline test/gtest/softmaxcrossentropywithlogits.hpp]: If it is not used, please delete it.

File: src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    size_t num_class = input_tv.size[1];
+
+    size_t batch_offset = gid * num_class;
+
+    // __local FSTYPE lmax[LOCAL_SIZE], lsum[LOCAL_SIZE];
+    __shared__ FLOAT_ACCUM lmax[LOCAL_SIZE], lsum[LOCAL_SIZE];
+    lmax[lid] = -INFINITY;
+    lsum[lid] = 0.0f;
+
+    // __local DTYPE output_grad_val;
```
Comment (et16kr) [inline src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp]: Please remove comment.

**Review by et16kr — COMMENTED** [review]
please make analyze
```
/home/MIOpen/test/cpu_softmaxcrossentropywithlogits.hpp:64:34: warning: Invalid std::log() argument nr 1. The value is 0 but the valid values are '4.94066e-324:'. [invalidFunctionArg]
        float log_sum = std::log(sum);
                                 ^
/home/MIOpen/test/cpu_softmaxcrossentropywithlogits.hpp:57:22: note: Assignment 'sum=0.0f', assigned value is 0.0
        float sum = 0.0f;
                     ^
/home/MIOpen/test/cpu_softmaxcrossentropywithlogits.hpp:58:29: note: Assuming condition is false
        for(size_t i = 0; i < num_class; ++i)
                            ^
/home/MIOpen/test/cpu_softmaxcrossentropywithlogits.hpp:64:34: note: Invalid argument
        float log_sum = std::log(sum);
                                 ^
```

File: src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp (hunk 5 above)
Comment (hieule88) [reply src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp]: removed it

File: test/gtest/softmaxcrossentropywithlogits.hpp (hunk 4 above)
Comment (hieule88) [reply test/gtest/softmaxcrossentropywithlogits.hpp]: uncommented it

File: src/solver/softmaxcrossentropywithlogits/fwd_con_softmaxcrossentropywithlogits.cpp (hunk 3 above)
Comment (hieule88) [reply src/solver/softmaxcrossentropywithlogits/fwd_con_softmaxcrossentropywithlogits.cpp]: removed it

File: src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp (hunk 2 above)
Comment (hieule88) [reply src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp]: removed it

File: src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp (hunk 1 above)
Comment (hieule88) [reply src/kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp]: removed it

**Comment by hieule88** [comment]
> please make analyze
> 
> ```
> /home/MIOpen/test/cpu_softmaxcrossentropywithlogits.hpp:64:34: warning: Invalid std::log() argument nr 1. The value is 0 but the valid values are '4.94066e-324:'. [invalidFunctionArg]
>         float log_sum = std::log(sum);
>                                  ^
> /home/MIOpen/test/cpu_softmaxcrossentropywithlogits.hpp:57:22: note: Assignment 'sum=0.0f', assigned value is 0.0
>         float sum = 0.0f;
>                      ^
> /home/MIOpen/test/cpu_softmaxcrossentropywithlogits.hpp:58:29: note: Assuming condition is false
>         for(size_t i = 0; i < num_class; ++i)
>                             ^
> /home/MIOpen/test/cpu_softmaxcrossentropywithlogits.hpp:64:34: note: Invalid argument
>         float log_sum = std::log(sum);
>                                  ^
> ```

I made analyze and fixed the warning by adding
```
float log_val; 
if(sum != 0) 
   log_val = std::log(sum); 
else 
   log_val = std::log(std::numeric_limits<float>::epsilon()); 
```

File: src/include/miopen/softmaxcrossentropywithlogits/problem_description.hpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        if(inputDesc.GetLengths()[0] != outputGradDesc.GetLengths()[0])
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "SoftmaxCrossEntropyWithLogits: Tensor sizes do not match.");
+        }
+        for(int i = 0; i < inputDesc.GetSize(); ++i)
+        {
+            if(inputDesc.GetLengths()[i] != backpropDesc.GetLengths()[i] ||
+               inputDesc.GetLengths()[i] != inputGradDesc.GetLengths()[i] ||
+               inputDesc.GetLengths()[i] != targetGradDesc.GetLengths()[i])
```
Comment (et16kr) [inline src/include/miopen/softmaxcrossentropywithlogits/problem_description.hpp]: Looking at the backend kernel, input_grad and target_grad are optional.
If so, the dims of those tensors may be empty() in some cases.

File: src/include/miopen/softmaxcrossentropywithlogits/problem_description.hpp [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "SoftmaxCrossEntropyWithLogits: Tensor sizes do not match.");
+            }
+        }
+        if(backpropDesc.GetSize() > 2 || inputDesc.GetSize() > 2 || inputGradDesc.GetSize() > 2 ||
+           targetGradDesc.GetSize() > 2)
```
Comment (et16kr) [inline src/include/miopen/softmaxcrossentropywithlogits/problem_description.hpp]: Isn't the tensor's dim equal to 2? Is even 1 possible?
And it would be better to check the dim size of the tensor first.

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -470,6 +475,7 @@ if( MIOPEN_BACKEND MATCHES "OpenCL" OR MIOPEN_BACKEND STREQUAL "HIPOC" OR MIOPEN
         kernels/MIOpenSoftmax.cl
         kernels/MIOpenSoftmaxAttn.cpp
         kernels/MIOpenSum.cpp
+        kernels/MIOpenSoftmaxCrossEntropyWithLogits.cpp
```
Comment (et16kr) [inline src/CMakeLists.txt:477]: Please move it up one line.

**Review by et16kr — COMMENTED** [review]
It is not common to calculate the grad of target from loss backward.
So, it would be a good idea to optionally include target grad in the driver.
However, there is no difference in performance even without target grad.
Or, if target grad calculation is common in SoftmaxCrossEntropyWithLogits
Please ignore my suggestion.

File: test/gtest/softmaxcrossentropywithlogits.cpp [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+                         testing::ValuesIn(SoftmaxCrossEntropyWithLogitsTestConfigs()));
+INSTANTIATE_TEST_SUITE_P(SoftmaxCrossEntropyWithLogitsTestSet,
+                         SoftmaxCrossEntropyWithLogitsTestBFloat16,
+                         testing::ValuesIn(SoftmaxCrossEntropyWithLogitsTestConfigs()));
+
+// BACKWARD TEST
+TEST_P(SoftmaxCrossEntropyWithLogitsTestFloatBwd, SoftmaxCrossEntropyWithLogitsTestBwd)
+{
+    if((miopen::IsEnabled(ENV(MIOPEN_TEST_ALL)) && GetFloatArg() == "--float") ||
+       GetFloatArg() == "--testall")
```
Comment (et16kr) [inline test/gtest/softmaxcrossentropywithlogits.cpp]: AMD said test all under the following conditions:
```miopen::IsUnset(ENV(MIOPEN_TEST_ALL))```
https://github.com/moreh-dev/MIOpen/pull/27#discussion_r1634984024

File: src/CMakeLists.txt (hunk 8 above)
Comment (hieule88) [reply src/CMakeLists.txt:477]: moved it in [ec341b9](https://github.com/moreh-dev/MIOpen/pull/30/commits/ec341b9a68e182cc9b96318816f190ae3d23de3b)

File: src/include/miopen/softmaxcrossentropywithlogits/problem_description.hpp (hunk 7 above)
Comment (hieule88) [reply src/include/miopen/softmaxcrossentropywithlogits/problem_description.hpp]: fixed it in [ec341b9](https://github.com/moreh-dev/MIOpen/pull/30/commits/ec341b9a68e182cc9b96318816f190ae3d23de3b)

File: src/include/miopen/softmaxcrossentropywithlogits/problem_description.hpp (hunk 6 above)
Comment (hieule88) [reply src/include/miopen/softmaxcrossentropywithlogits/problem_description.hpp]: fixed it in [ec341b9](https://github.com/moreh-dev/MIOpen/pull/30/commits/ec341b9a68e182cc9b96318816f190ae3d23de3b)

File: test/gtest/softmaxcrossentropywithlogits.cpp (hunk 9 above)
Comment (hieule88) [reply test/gtest/softmaxcrossentropywithlogits.cpp]: removed it

**Comment by hieule88** [comment]
> However, there is no difference in performance even without target grad.
> Or, if target grad calculation is common in SoftmaxCrossEntropyWithLogits
> Please ignore my suggestion.

I made computing target grad optional by adding "-T" flag in driver in commit [ec341b9](https://github.com/moreh-dev/MIOpen/pull/30/commits/ec341b9a68e182cc9b96318816f190ae3d23de3b)

**Comment by et16kr** [comment]
I appreciate your effort. good job.
